test/test.js

```diff
@@ -89,12 +93,30 @@ describe('http client', () => {
       const response = await client.get('/get', { foo: 'bar' })
       assert.deepEqual(response, { hello: 'world' })
     })
+
+    it('should handle errors', async () => {
+      try {
```

Comment:
I was thinking we could also check for response data that we define in nock: `.reply(500, { hello: 'world' })`.

In that case, the test could look like this:

```javascript
it('should handle errors', async () => {
  await assert.rejects(
    client.get('/error'),
    { message: '{"hello":"world"}' }
  )
})
```
But that would require us to reject the promise with Error instead of data here:
https://github.com/elastic/enterprise-search-node/blob/e3a7c4c7f688c8572780bd8f143fbf93f23a7039/lib/client.js#L30
So error handling in that line could look like this:
```javascript
return reject(new Error(JSON.stringify(data)))
```

What do you think about this?

Reply:
I like the idea of using `assert.reject`!
I just checked and the test would change to this:
```js
    it('should handle errors', async () => {
      await assert.rejects(
        client.get('/error'),
        { hello: 'world' }
      )
    })

```

Regarding what to reject, it all depends on what you want to show to the user in case of error. Are we hiding the HTTP related details?
If so something like the following would work:
```js
  get(path, params) {
    return new Promise((resolve, reject) => {
      sget({ ... }, (err, res, data) => {
        if (err || res.statusCode >= 400) {
			err = err || new Error('Something went wrong')
			err.payload = data
			return reject(err)
		}
        resolve(data)
      })
    })
  }
```

Otherwise, if we are ok to expose also the HTTP related parameters, we can do this:
```js
  get(path, params) {
    return new Promise((resolve, reject) => {
      sget({ ... }, (err, res, data) => {
        if (err || res.statusCode >= 400) {
			// here the message could be the status code message
			err = err || new Error('Something went wrong')
			err.body = data
			err.statusCode = res.statusCode
			err.headers = res.headers
			return reject(err)
		}
        resolve(data)
      })
    })
  }
```

While the first solution is more concise, the second may help to provide a better debugging experience.

Reply:
Great suggestion! Let's show more rather than less (2nd option). And I like the idea of using status code in error message.

1 question though: If we get an error `err`, will we have access to `data`, `res.statusCode` and `res.headers`?

---
Note: I updated the PR branch with the latest master. The only thing that has changed is we dropped support for Node.js 8. After pulling the branch, run `npm use` before running the tests.

Reply:
> 1 question though: If we get an error err, will we have access to data, res.statusCode and res.headers?

It depends on the error, if it's a connection reset or a timeout, we will not have HTTP related details.

Reply:
I've updated the code with the new enhanced errors :)
There is also a new test to show what happens in case of a not HTTP related error.

Reply:
This looks great!
I like the new test and the usage of http.STATUS_CODES 👍